--- 1.4.2.7/RRI.f90	2023-11-22 09:46:40.000000000 +0900
+++ 1.4.2.7/RRI_UNST.f90	2025-12-25 00:00:00.000000000 +0900
@@ -6,6 +6,13 @@
 !
 program RRI
 use globals
+use unst_globals_mod
+use unst_write_procedures
+use unst_wrfile
+use unst_prewrfile
+use unst_read
+use d1riv_globals_mod
+use unst_d1riv_read
 use runge_mod
 use dam_mod, only: dam_switch, dam_vol_temp
 use tecout_mod
@@ -66,6 +73,9 @@
 character*256 ofile_ro, outdir1, outdir2
 integer kk, l
 
+integer uflg
+uflg = 0
+
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
 !!! STEP 0: FILE NAME AND PARAMETER SETTING
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
@@ -506,6 +516,7 @@
   read(11, *, iostat = ios) (rdummy, j = 1, nx_rain)
  enddo
  if( ios.lt.0 ) exit
+ if( ios.ne.0 ) exit
  tt = tt + 1
 enddo
 tt_max_rain = tt - 1
@@ -526,6 +537,8 @@
 ! unit convert from (mm/h) to (m/s)
 qp = qp / 3600.d0 / 1000.d0
 
+dtrain = t_rain(1) - t_rain(0)
+
 do j = 1, nx
  rain_j(j) = int( (xllcorner + (dble(j) - 0.5d0) * cellsize - xllcorner_rain) / cellsize_rain_x ) + 1
 enddo
@@ -575,6 +588,27 @@
  close(11)
 endif
 
+call unst_rdat(ny_rain, qp, nx_rain, tt_max_rain, &
+    ny, nx, lasth, dt, &
+    xllcorner_rain, yllcorner_rain, cellsize_rain_x, cellsize_rain_y, &
+    xllcorner, yllcorner, cellsize, dir)
+call open_unst_output_files
+if(dsmesh==1) call dsmeshdat(lasth)
+if(d1riv==1) call d1rivdat(lasth, dt2, mesh, baseo, fd1riv_cntl)
+if(plantFN==1) call plantFNdat
+if(plantDa==1) call plantDadat
+if(paddydam==1) call paddydat
+if(drainarea==1) call draindat
+if(morid==1) call moriddat
+
+call unst_initiald(dir, nx, ny)
+if(d1riv==1) call d1rivinitiald(dt2)
+if(paddydam==1) call paddyinitiald
+if(drainarea==1) call draininitiald
+
+call diskwrite
+call dispwrite
+
 ! For TSAS Output (Initial Condition)
 call sub_slo_ij2idx( hs, hs_idx )
 call sub_riv_ij2idx( hr, hr_idx )
@@ -593,6 +627,7 @@
 out_dt = dble(maxt) / dble(outnum)
 out_dt = max(1.d0, out_dt)
 out_next = nint(out_dt)
+timmax = dble(dt) * dble(calldt)
 tt = 0
 
 do t = 1, maxt
@@ -1193,6 +1228,17 @@
   endif
  endif
 
+ call unst_qin(ny, nx, i4, qr_ave, qs_ave, hr, hs, area, time, uflg)
+ if(mod(time, timmax) == 0 .and. uflg==0) call predispwrite(time)
+ if(mod(time, timmax) == 0 .and. uflg==0) call prediskwrite(time)
+ if(mod(time, timmax) == 0 .and. uflg==1) then
+  call UNST2D(ny, nx, domain, riv, time, hs, hr)
+
+  ! calculate river and slope interaction
+  call funcrs(hr, hs)
+  write(*,*) 'UNST2D >>> RRI h replaced'
+ endif
+
  ! check water balance
  if(mod(t, 1).eq.0) then
   call storage_calc(hs, hr, hg, ss, sr, si, sg)
@@ -1204,6 +1250,39 @@
 
 enddo
 
+call wrhmax
+
+write(*, 1999) time
+1999 format('      - normal end -  time=', f8.0)
+
+call close_unst_output_files
+
+deallocate(baseo, dnox, dnoy, smesh, scv, rthl, ux, uy, xmesh, ymesh, rtuv_x, rtuv_y)
+deallocate(limesh, linode, inf, ko, menode, melink, inl, qin, lkyokai_dx, lkyokai_dy)
+deallocate(unsth, ho, hl, hmax, uummax, vvmmax)
+deallocate(um, umo, umm, uu, vn, vno, vnm, vv)
+deallocate(mn, rnof, lambda, rbeta)
+deallocate(uum, vvm, lhan, lhano, qr_sum, rnx, dl)
+if(plantFN==1) deallocate(plantF_array, plantN_array)
+if(plantDa==1) deallocate(dk_val)
+if(paddydam==1) deallocate(paddyid, pqout_idx, pdrain, min_pmeshid, device)
+if(paddydam==1) deallocate(orifice_num, min_dist, psmesh, dr_dist, dhp, phid)
+if(paddydam==1) deallocate(paddy_q, pqh, drain2phidx)
+if(drainarea==1) deallocate(inf_dr, drp, drr, drr_dist, dhj)
+if(dsmesh==1) then
+    if(any(ds_inf==2)) deallocate(ds_wl)
+    if(dsmesh==1) deallocate(ds_dt, ds_inf, ds2me, ds_upme)
+endif
+if(d1riv==1) deallocate(riv_ndan, h_1d, ho_1d, vv_1d, q_1d, qo_1d, a_1d, r_1d, b_1d, rn_1d)
+if(d1riv==1) deallocate(h_1dmax, ntype_1d, kp_1d, dx_1d, depth_1d, rbed_1d, rzmax_1d)
+if(d1riv==1) deallocate(dan_record, n_tbl, h_table, a_table, r_table, b_table, rn_table)
+if(d1riv==1) deallocate(b_idx_1d, b_dt_1d, b_data_1d, up_h_1d, up_q_1d, b_upme_1d, b_dome_1d)
+if(d1riv==1) deallocate(subflow_input, subq_all, tributaryq_1d, pumpq_1d, sluiceq_1d)
+if(d1riv==1) deallocate(breakq_1d, breakq_l_1d, breakq_r_1d)
+if(d1riv==1) deallocate(weirq_1d, weirq_l_1d, weirq_r_1d, weir_dx_1d)
+if(d1riv==1) deallocate(w_alpha_1d, w_angle_1d, bktype_1d)
+if(d1riv==1) deallocate(ncnct_1d2d, cnct_1d2d_idx, crown_1d, inland_1d)
+
 !pause
 
 end program RRI
